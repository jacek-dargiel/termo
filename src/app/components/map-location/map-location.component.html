<div
  class="temperature-box"
  *ngIf="!loading; else spinner"
>
  <header>
    <h2>ðŸŒ¡<span class="label">{{location.name}}</span></h2>
  </header>
  <div class="temperatures">
    <div class="temperature temperature--primary">
      <span class="value">{{location.lastMeasurmentValue | toFixed}}</span><span class="unit">Â°C</span>
    </div>
    <div
      class="temperature temperature--minimal"
      *ngIf="location.minimalMeasurmentValue !== null"
    >
      <span class="value">Min: {{location.minimalMeasurmentValue | toFixed}}</span><span class="unit">Â°C</span>
    </div>
  </div>
  <div
    class="footer footer--warning footer--warning-outdated"
    *ngIf="location.updatedAt | isLocationOutdated"
  >
    <span>OpÃ³Åºnione: {{ location.updatedAt | relativeTime }}</span>
  </div>
</div>
<div class="arrow"></div>
<ng-template #spinner>
  <termo-spinner></termo-spinner>
</ng-template>
