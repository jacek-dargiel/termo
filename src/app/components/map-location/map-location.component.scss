@import '../../../styles/variables';

$activeColor: #eede25;
$mutedColor: mix(#777, $colorPrimary, 80%);

:host {
  position: absolute;
  background: #fff;
  outline: 1px solid mix(#bbb, $colorPrimary, 85%);
  box-shadow: 0.3rem 0.3rem 1rem rgba(#000, 0.5);
  padding: 0.4rem 1rem;
  overflow: hidden;
  width: 13rem;
  // margin-left: -13rem;
  text-align: center;
  display: block;
  cursor: pointer;

  &.location--selected {
    border-bottom: 0.2rem solid $colorPrimary;
  }

  &:focus,
  &:active {
    outline: 0.2rem solid $activeColor;
    user-select: none;

    .arrow {
      background: $activeColor;
    }
  }

  &:hover {
    h2 .label{
      text-decoration: underline;
      text-decoration-color: rgba($mutedColor, 0.5);
    }
  }
  
  @media (max-width: $breakpointMobile) {
    box-sizing: border-box;
    position: relative;
    margin: 1rem auto;
    width: calc(100% - 2rem);
    bottom: unset !important;
    right: unset !important;
  }
}

.arrow {
  $size: 15px;
  position: absolute;
  width: $size;
  height: $size;
  bottom: -1*$size/2;
  right: -1*$size/2;
  background: $colorPrimary;
  transform: rotate(0.125turn);

  @media (max-width: $breakpointMobile) {
    display: none;
  }
}

header h2 {
  margin: 0;
  font-size: 1rem;
  text-transform: uppercase;
  color: $mutedColor;
}

.temperature{
  font-weight: 600;
  margin-bottom: 0.5rem;

  &--primary {
    color: $colorPrimary;
    font-size: 3rem;
  }
  &--minimal {
    border-top: 1px solid mix(#555, $colorPrimary, 80%);
    width: 9rem;
    margin: 0.2rem auto 0.5rem auto;
    padding-top: 0.2rem;
    color: #555;
  }
}

.footer {
  font-size: 0.88rem;
  color: $mutedColor;
  
  &--warning {
    &::before {
      content: '⚠️'
    }
  }
}
