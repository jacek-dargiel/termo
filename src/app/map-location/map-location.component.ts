import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'termo-map-location',
  templateUrl: './map-location.component.html',
  styleUrls: ['./map-location.component.css']
})
export class MapLocationComponent implements OnInit {

  constructor() { }

  ngOnInit() {
  }

}
