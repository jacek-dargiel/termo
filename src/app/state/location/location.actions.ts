import { Action } from '@ngrx/store';
import { Location } from './location.model';

export enum LocationActionTypes {
  MapInitialized = '[Map] Map Initialized',
  RefreshMeasurmentsStart = '[Effect] Refresh Measurments Start',
  RefreshMeasurmentsFinish = '[Effect] Refresh Measurments Finish',
  FetchLocationsSuccess = '[API] Fetch Locations Success',
  FetchLocationsError = '[API] Fetch Locations Error',

  SelectLocation = '[Map] Select Location',
}

export class FetchLocationsSuccess implements Action {
  readonly type = LocationActionTypes.FetchLocationsSuccess;

  constructor(public payload: { locations: Location[] }) {}
}

// ### Custom actions

export class MapInitialized implements Action {
  readonly type = LocationActionTypes.MapInitialized;
}

export class RefreshMeasurmentsStart implements Action {
  readonly type = LocationActionTypes.RefreshMeasurmentsStart;
  constructor(public payload: { location: Location }) {}
}

export class RefreshMeasurmentsFinish implements Action {
  readonly type = LocationActionTypes.RefreshMeasurmentsFinish;
}

export class FetchLocationsError implements Action {
  readonly type = LocationActionTypes.FetchLocationsError;
  constructor(public payload: { error: Error }) {}
}

export class SelectLocation implements Action {
  readonly type = LocationActionTypes.SelectLocation;
  constructor(public payload: { location: Location }) {}
}

export type LocationActions =
 FetchLocationsSuccess
 | MapInitialized
 | RefreshMeasurmentsStart
 | RefreshMeasurmentsFinish
 | FetchLocationsError
 | SelectLocation
 ;
