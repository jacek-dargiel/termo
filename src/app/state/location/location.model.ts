import { Point } from '../../interfaces';

export interface Location {
  id: number;
  name: string;
  mapPosition: Point;
  feedKey: string;
  updatedAt: Date;
}
