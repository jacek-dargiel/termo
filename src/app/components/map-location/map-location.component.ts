import { Component, Input, HostBinding, OnInit } from '@angular/core';
import { LocationWithKeyMeasurmentValues } from '../../state/location/location.model';

@Component({
  selector: 'termo-map-location',
  templateUrl: './map-location.component.html',
  styleUrls: ['./map-location.component.scss']
})
export class MapLocationComponent implements OnInit {
  @Input() location: LocationWithKeyMeasurmentValues;
  @Input() loading: boolean;
  @HostBinding('style.top.%') top;
  @HostBinding('style.left.%') left;

  constructor(
  ) { }

  ngOnInit() {
    this.adjustPosition();
  }

  adjustPosition() {
    this.left = this.location.mapPosition.x * 100;
    this.top = this.location.mapPosition.y * 100;
  }

}
