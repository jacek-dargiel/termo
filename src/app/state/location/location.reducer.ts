import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { Location } from './location.model';
import { LocationActions, LocationActionTypes } from './location.actions';
import { MeasurmentActionTypes, MeasurmentActions } from '../measurment/measurment.actions';
import { Measurment } from '../measurment/measurment.model';
import { compareAsc } from 'date-fns/esm';
import { Dictionary } from '@ngrx/entity/src/models';
import { last } from '../../helpers/lodash';

export interface State extends EntityState<Location> {
  loading: boolean;
  latestMeasurmentIDs: Dictionary<string>;
  selected: string;
}

function locationPositionComparer(a: Location, b: Location) {
  let aOffset = a.mapPosition.x + a.mapPosition.y;
  let bOffset = b.mapPosition.x + b.mapPosition.y;

  return aOffset - bOffset;
}

export let adapter: EntityAdapter<Location> = createEntityAdapter<Location>({
  sortComparer: locationPositionComparer,
});

export const INITIAL_STATE: State = adapter.getInitialState({
  loading: false,
  latestMeasurmentIDs: {},
  selected: undefined,
});

export function reducer(
  state = INITIAL_STATE,
  action: LocationActions | MeasurmentActions
): State {
  switch (action.type) {

    case LocationActionTypes.FetchLocationsSuccess: {
      let loadedState = {
        ...state,
      };
      return adapter.addAll(action.payload.locations, loadedState);
    }

    case LocationActionTypes.MapInitialized: {
      return {
        ...state,
        loading: true,
      };
    }

    case LocationActionTypes.RefreshButtonClick:
    case LocationActionTypes.FetchLocationsSuccess:
    case LocationActionTypes.RefreshSignal:
    {
      return {
        ...state,
        loading: true,
      };
    }

    case MeasurmentActionTypes.FetchMeasurmentsError: {
      return {
        ...state,
        loading: false,
      };
    }

    case LocationActionTypes.FetchLocationsError: {
      return {
        ...state,
        loading: false,
      };
    }

    case LocationActionTypes.RefreshMeasurmentsFinished: {
      return {
        ...state,
        loading: false,
      };
    }

    case MeasurmentActionTypes.FetchMeasurmentsSuccess: {
      let locationsState = state;
      if (action.payload.measurments.length > 0) {
        let sortedMeasurments = sortMeasurments(action.payload.measurments);
        let latestMeasurment = last(sortedMeasurments);
        locationsState = adapter.updateOne({id: action.payload.locationId, changes: { updatedAt: latestMeasurment.created_at }}, state);
        locationsState = {
          ...locationsState,
          latestMeasurmentIDs: {
            ...locationsState.latestMeasurmentIDs,
            [action.payload.locationId]: latestMeasurment.id,
          }
        };
      }

      return {
        ...locationsState,
      };
    }

    case LocationActionTypes.SelectLocation: {
      return {
        ...state,
        selected: action.payload.location && action.payload.location.id,
      };
    }

    default: {
      return state;
    }
  }
}

export const {
  selectIds,
  selectEntities,
  selectAll,
  selectTotal,
} = adapter.getSelectors();

function sortMeasurments(measurments: Measurment[]) {
  return [...measurments].sort((a: Measurment, b: Measurment) => compareAsc(a.created_at, b.created_at));
}
