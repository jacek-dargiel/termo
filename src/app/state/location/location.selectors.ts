import * as fromLocation from './location.reducer';
import { createFeatureSelector, createSelector } from '@ngrx/store';

export const selectLocationState = createFeatureSelector<fromLocation.State>('location');

export const selectLoading = createSelector(
  selectLocationState,
  location => location.loading
);

export const selectAllLocations = createSelector(
  selectLocationState,
  fromLocation.selectAll,
);
