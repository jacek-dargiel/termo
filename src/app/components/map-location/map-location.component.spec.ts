import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';

import { MapLocationComponent } from './map-location.component';
import { ToFixedPipe } from '../../pipes/to-fixed.pipe';
import { IsLocationOutdatedPipe } from '../../pipes/is-location-outdated.pipe';
import { RelativeTimePipe } from '../../pipes/relative-time.pipe';
import { LocationWithKeyMeasurmentValues } from '../../state/location/location.model';
import { TERMO_CURRENT_TIME_FACTORY } from '../../pipes/current-time.injection-token';

describe('MapLocationComponent', () => {
  let component: MapLocationComponent;
  let fixture: ComponentFixture<MapLocationComponent>;

  beforeEach(waitForAsync(() => {
    let since = new Date('2018-09-19T22:15:00');
    TestBed.configureTestingModule({
      declarations: [
        MapLocationComponent,
        ToFixedPipe,
        IsLocationOutdatedPipe,
        RelativeTimePipe,
      ],
      providers: [
        { provide: TERMO_CURRENT_TIME_FACTORY, useValue: () => since },
      ],
      schemas: [NO_ERRORS_SCHEMA]
    })
    .compileComponents();
  }));

  let baseLocation: LocationWithKeyMeasurmentValues;
  beforeEach(() => {
    fixture = TestBed.createComponent(MapLocationComponent);
    component = fixture.componentInstance;
    baseLocation = {
      id: 'temperatura.gorny_tunel',
      name: 'GÃ³rny Tunel',
      mapPosition: {
        x: 0.25,
        y: 0.75,
      },
      updatedAt: new Date('2018-09-19T22:10:00'),
      lastMeasurmentValue: 21.12,
      minimalMeasurmentValue: 19.5,
    };
    component.location = baseLocation;

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render loading state', () => {
    component.loading = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should properly render base state', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should display outdated warning', () => {
    let testLocation = {
      ...baseLocation,
      updatedAt: new Date('2018-09-15T22:10:00'),
    };
    component.location = testLocation;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should render selected state', () => {
    component.selected = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should position itself on the map element', () => {
    fixture.detectChanges();
    expect(fixture.debugElement.styles.bottom).toBe('25%');
    expect(fixture.debugElement.styles.right).toBe('75%');
  });

  it('should emit `select` event on click', () => {
    let fakeSelectOutput = jest.spyOn(component.select, 'emit');
    fixture.debugElement.nativeElement.click();
    expect(fakeSelectOutput).toHaveBeenCalled();
  });
});
