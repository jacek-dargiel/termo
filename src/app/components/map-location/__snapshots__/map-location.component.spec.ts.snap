// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapLocationComponent should display outdated warning 1`] = `
<termo-map-location
  bottom={[Function Number]}
  location={[Function Object]}
  right={[Function Number]}
  select={[Function EventEmitter]}
>
  <div
    class="temperature-box"
  >
    <header>
      <h2>
        ðŸŒ¡
        <span
          class="label"
        >
          GÃ³rny Tunel
        </span>
      </h2>
    </header>
    <div
      class="temperatures"
    >
      <div
        class="temperature temperature--primary"
      >
        <span
          class="value"
        >
          21.12
        </span>
        <span
          class="unit"
        >
          Â°C
        </span>
      </div>
      
      <div
        class="temperature temperature--minimal"
      >
        <span
          class="value"
        >
          Min: 19.50
        </span>
        <span
          class="unit"
        >
          Â°C
        </span>
      </div>
    </div>
    
    <div
      class="footer footer--warning footer--warning-outdated"
    >
      <span>
        OpÃ³Åºnione: 4 dni
      </span>
    </div>
  </div><div
    class="arrow"
  />
</termo-map-location>
`;

exports[`MapLocationComponent should properly render base state 1`] = `
<termo-map-location
  bottom={[Function Number]}
  location={[Function Object]}
  right={[Function Number]}
  select={[Function EventEmitter]}
>
  <div
    class="temperature-box"
  >
    <header>
      <h2>
        ðŸŒ¡
        <span
          class="label"
        >
          GÃ³rny Tunel
        </span>
      </h2>
    </header>
    <div
      class="temperatures"
    >
      <div
        class="temperature temperature--primary"
      >
        <span
          class="value"
        >
          21.12
        </span>
        <span
          class="unit"
        >
          Â°C
        </span>
      </div>
      
      <div
        class="temperature temperature--minimal"
      >
        <span
          class="value"
        >
          Min: 19.50
        </span>
        <span
          class="unit"
        >
          Â°C
        </span>
      </div>
    </div>
    
  </div><div
    class="arrow"
  />
</termo-map-location>
`;

exports[`MapLocationComponent should render loading state 1`] = `
<termo-map-location
  bottom={[Function Number]}
  loading={[Function Boolean]}
  location={[Function Object]}
  right={[Function Number]}
  select={[Function EventEmitter]}
>
  <termo-spinner /><div
    class="arrow"
  />
</termo-map-location>
`;

exports[`MapLocationComponent should render selected state 1`] = `
<termo-map-location
  bottom={[Function Number]}
  location={[Function Object]}
  right={[Function Number]}
  select={[Function EventEmitter]}
  selected={[Function Boolean]}
>
  <div
    class="temperature-box"
  >
    <header>
      <h2>
        ðŸŒ¡
        <span
          class="label"
        >
          GÃ³rny Tunel
        </span>
      </h2>
    </header>
    <div
      class="temperatures"
    >
      <div
        class="temperature temperature--primary"
      >
        <span
          class="value"
        >
          21.12
        </span>
        <span
          class="unit"
        >
          Â°C
        </span>
      </div>
      
      <div
        class="temperature temperature--minimal"
      >
        <span
          class="value"
        >
          Min: 19.50
        </span>
        <span
          class="unit"
        >
          Â°C
        </span>
      </div>
    </div>
    
  </div><div
    class="arrow"
  />
</termo-map-location>
`;
